parser/src/format/mod.rs

```diff
@@ -253,6 +255,102 @@ fn compression_from_filename(filename: &str) -> Option<Compression> {
     })
 }
 
+/// A decorating iterator that tracks parsing errors and absorbs a specified
+/// rate of errors. If that rate is exceeded, then all the errors encountered
+/// are returned. Should not be polled again once an error is returned.
+#[derive(Debug)]
+struct ParseErrorBuffer<I> {
+    /// The iterator we're wrapping.
+    inner: I,
+    /// The amount of errors we can absorb before halting parsing.
+    threshold: ErrorThreshold,
+    /// The number of records we've seen.
+    total_records: usize,
+    /// The number of errors we've seen lately.
+    errors_in_latest_window: usize,
+    /// The error messages from all the rows that have failed to parse.
+    absorbed: Vec<ParseError>,
+}
+
+impl<I: Iterator> ParseErrorBuffer<I> {
+    /// Consumes the next item out of the inner iterator.
+    fn advance(&mut self) -> Option<I::Item> {
+        // How many of the recent records to consider when trying to decide if
+        // we've entered a new region of bad data of the file.
+        const ERROR_BUFFER_WINDOW_SIZE: usize = 1000;
+
+        self.total_records += 1;
+
+        if self.total_records > ERROR_BUFFER_WINDOW_SIZE {
```

Comment:
Hmm, yeah. I got too clever. :facepalm: Plus my test cases hid the problem. 

I'll take a crack at the rate limiter approach. I was trying to avoid having to store data per row processed. I think this will be a pretty similar to how it is implemented now, but hopefully less broken.

Reply:
Upon further reflection, the token bucket approach makes understanding the error detection quite opaque. The algorithm ended up being very dependent upon initial conditions, which makes it rather difficult to understand if it was working correctly.

As discussed, now we just track the exact number of errors within the window at any given point and fail if we have seen too many errors.